### PR TITLE
CLI to assume namespace from cwd #19

### PR DESCRIPTION
**Pull request by Seke1412**
1. If no namespace indicate => check for app dir existed
    1.1 No app dir existed => create app/components/[name]
    1.2 App dir existed => create components/[name]
2. with namespace indicate => start from current dir

@mikabytes  during implementation, i found some edge case about create component outside app directory, so i've some assumption base on the previous code, whole logic:
```
// path given: east/asia
// cwd: app
// => app/east/asia/components/test
if (havePath && haveAppDir)  {
    resolvePath = path
}


// path given: ''
// cwd: 'cannot find app directory in absolute path by cwd'
// => app/components/test
if (!havePath && !haveAppDir)  {
    resolvePath = `app`
}

// path given: ''
// cwd: app/east/asia
// => app/east/asia/components/test
if (!havePath && haveAppDir)  {
    resolvePath = ``
}

// path given: 'east/asia'
// cwd: 'cannot find app directory in absolute path by cwd'
// => app/east/asia/components/test
if (havePath && !haveAppDir)  {
    resolvePath = `app/${path}/`
}
```

test case included, i still a little confuse on those edge cases, so if you found it reasonable i will apply to other create commands

### CODE REVIEW

**Comment by mikabytes** [comment]
Hey, thanks for this!

It looks very reasonable. I have more time to look at it tomorrow to be completely sure.

One thought. Maybe it would be easier to do this logic one level higher up (i.e. before the createComponent command). That way it would work for all commands automatically. But, just an idea. It is late :)

**Comment by Seke1412** [comment]
@mikabytes actually i also found this implementation quite confuse. I also like the idea of taking all the parent directory's name as namespace except app and root, not sure i'm in the same page as yours suggestion, i will list out the case here to make sure we're aligned:

case 1:
- cwd: `app/east/asia/`
- cmd: create component test 
- expected dir: app/east/asia/components/tests/.... 
- expected namespace: `x-east-asia-test`  (please also consider we can have south/asia, if we just take `x-asia-test` only, then they will be conflict => not sure on this part, really need your comment @mikabytes )

case 2:
- cwd: `app/east/asia/`
- cmd: create component test **abc**
- expected dir: app/east/asia/**abc**/components/tests/.... 
- expected namespace:  `x-east-asia-abc-test`

case 3:
- cwd: `east/asia/` (create component somewhere outside app folder)
- cmd: create component test
- expected dir: app/east/asia/components/tests/.... 
- expected namespace:  `x-global-east-asia-test`

case 4:
- cwd: `east/asia/` (create component somewhere outside app folder)
- cmd: create component test **abc**
- expected dir: east/asia/**abc**/components/tests/.... 
- expected namespace: `x-global-east-asia-abc-test`

case 5:
- cwd: `/` (create component somewhere outside app folder)
- cmd: create component test **[abc]**
- expected dir: /**[abc/]**components/tests/.... 
- expected namespace: `x-global-[abc-]test`

**Comment by mikabytes** [comment]
@Seke1412 Man, I really love it how you present these scenarios. It is really helpful for me to give feedback :+1: 

Case 1: ✅ Yes, this is correct
Case 2: ✅ Yes, this is correct 
Case 3: ❌ Hmm, I think the command should fail here. It doesn't make sense to create things [1] outside of the `app` folder. 
Case 4: ❌ Same as case 3
Case 5: ❔ We'll have a little exception when in the project root folder (CWD `/`) because most often commands will be executed here. So we assume `app`. The namespace should not include the word `global`
Case 6: CWD: `..`: Making commands outside of a Tiden project is forbidden.

Notes:
- Adding `global`  namespace should not be needed anywhere. `app` is the global namespace :)

[1] "things" here meaning all of components, streams, pages, and nanos

**Comment by mikabytes** [comment]
Actually, one way of thinking about it is that `x` actually means `app`.

**Comment by mikabytes** [comment]
If we can do this "translation" of namespace higher up so that they are applied to _all_ commands, then that would be beautiful :)

Because, it does make sense to assume namespace from cwd for all of pages, nanos, streams, and components

**Comment by Seke1412** [comment]
I got your idea now, let me start with `create` command first, all good then i will move to the others. Cheers

**Comment by Seke1412** [comment]
@mikabytes  i've push new commit, when we've aligned the solution, i will apply to other command 👍

**Comment by mikabytes** [comment]
Cool! I hope to have some time to dog into this tastiness tomorrow.

File: ops/createComponent.test.js [hunk 1]
```diff
@@ -29,17 +29,32 @@ describe(`createComponent`, () => {
 
     it(`should throw an error`, async () => {
       try {
-        await createComponent({ path: `one`, name: `myComponent` })
+        await createComponent({ path: `app/one`, name: `myComponent` })
```
Comment (mikabytes) [inline ops/createComponent.test.js:32]: `app` should be implicitly understood. As everything is always in `app`, I don't want to have to write `app/something` all the time. `something` is quite enough.

So I think we should change this test back to the way it was before. All commands are always about `app`.

However, if you are currently CWD to a directory that is not root and not in `app` folder, then it should fail.

File: ops/createComponent.test.js [hunk 2]
```diff
@@ -29,17 +29,32 @@ describe(`createComponent`, () => {
...
       }
     })
   })
 
+  describe(`when create outside app dir`, () => {
+    beforeEach(async () => {
+      await mkdirp(`app/one/two/three`)
+    })
+
+    it(`should throw an error`, async () => {
```
Comment (mikabytes) [inline ops/createComponent.test.js:45]: When in root folder it should assume `app`.

Just as a convenience thing. Because most commands will be executed in root folder.

File: ops/createComponent.test.js [hunk 3]
```diff
@@ -113,6 +183,7 @@ describe(`createComponent`, () => {
   describe(`with args`, () => {
     beforeEach(async () => {
       await createComponent({
+        path: `app`,
```
Comment (mikabytes) [inline ops/createComponent.test.js:186]: path should be optional, defaulting to empty string

**Review by mikabytes — CHANGES_REQUESTED** [review]
This is getting a bit complicated. Sorry about that. The cases we outlined in the conversation are correct, but this implementation is now requiring the user to specify `app` before each command. That is not correct.

So, to clarify:

1. `cwd .`: namespace is `app`
2. `cwd app/`: namespace is `app`
3. `cwd lib/`: error, you can't create a component outside of app (root folder being an exception [1] where app is assumed).
4. `cwd app/burgers`: namespace is `burgers`
5. `cwd app/burgers/ketchup`: namespace is `burgers/ketchup`
6. `cwd ..`: error, can't create outside of project root.

**Comment by mikabytes** [comment]
Also, I just realized this has to be one level up (in CLI files, not in OPS) because OPS should be usable without a command line (it will integrate with other tools in the future). So we don't have access to `cwd` there.

So there's an additional constraint to this issue:
- Only files `cli.js` and files in `cli/` folder may be modified.

Let me know if this becomes a headache for you. I think you can do it. But if it becomes too much of a pain for you, then I will step in of course :)

**Comment by mikabytes** [comment]
Also, kudos for using [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) :)

**Comment by Seke1412** [comment]
@mikabytes , I realize that this issue should better be solved in `cli` level rather than command function level, i will close it. I will try another PR with `cli` level instead on Wednesday, if you have any rush on this issue, don't hesitate to step-in man (don't get me wrong, just in case something related to business, things should be done asap, i totally get it). Solving such kind of problem also teach me a lot, really appreciate your comments ;)

**Comment by mikabytes** [comment]
There's no rush, take all the time you need. I just hope you're enjoying the work, your contributions are much appreciated :)

The only deadline we have is to finish the current milestone (Stone) as soon as possible. But there are many other, more time-consuming, tasks needed for that which is taking my attention. So there's plenty of time for this issue.

File: ops/createComponent.js [hunk 4]
```diff
@@ -8,33 +8,36 @@ import camelToSnake from "../lib/camelToSnake.js"
...
 }) {
-  const realPath = path ? `app/${path}` : `app`
-  path = path || ``
-
-  const file = `${realPath}/components/${name}.js`
-  const demo = `${realPath}/components/${name}/demo.js`
-  const cssFile = `${realPath}/components/${name}/css.js`
+  const isInApp = process.cwd().indexOf(`/app`) !== -1
+  const isAppPath = path?.startsWith(`app`)
+  const resolvePath = (isInApp || isAppPath) ? path : `app/` + path
```
Comment (Seke1412) [inline ops/createComponent.js:20]: @mikabytes maybe i need you to jump-in here, it's not about solution, but i cannot make a decision here.
The Solution:
- I just need move those 3 lines of code to `cli` level, which mean everything related to path resolved, path transform will be handled in `cli` => create command level will receive whatever path cli pass into, execute the rest of logic (create component/css/imports/...)

The Issue:
- All of test case of `ops` and `cli` need to be rewrite, because:
- `cli/parse.test.js` expect object change (we already resolve the `path` with prefix `app`)
- `ops/*.text.js` now test case must give resolved path to create methods, create method itself doesn't resolve path anymore (as solution given)

Example:
- Given: 
   + cwd: "./"
   + command: "tiden create component myComponent one"
- Expected:
   + `cli` check that namespace doesn't point to app path, cwd not in app dir also => implicit resolve namespace to `app/one`
   + `ops` receive namespace as `app/one`, know nothing about what `cli` did
- Issue: all test case in `ops/createComponents.test.js` fail, due to calling direct to createComponent methods instead of go through cli

Comment (mikabytes) [reply ops/createComponent.js:20]: This issue is becoming more complicated for every day, haha :) But it's good, your investigations are leading us forward to better arch.

I will need to think deeply about this. But a few initial impressions:

1. `cli/parse.js` should only ever parse. Nothing else. So we cannot make the "translation" of namespace directory here.
2. Anything `ops` should receive the full details needed to do their work. They should not do modifications based on `cwd`.

So it seems there might be a need for a 3rd layer somewhere. I'll spend some time trying to figure this out now.

Comment (Seke1412) [reply ops/createComponent.js:20]: > I will need to think deeply about this. But a few initial impressions:
> 
> 1. cli/parse.js should only ever parse. Nothing else. So we cannot make the "translation" of namespace directory here.
> 2. Anything ops should receive the full details needed to do their work. They should not do modifications based on cwd.
> 

couldn't agree more, right here waiting for you my man

Comment (mikabytes) [reply ops/createComponent.js:20]: Moving this conversation to https://github.com/tidenjs/tiden/issues/19
